# COMP 641 Group Programming Project    

**Project Title:** Predicting 30-Day Readmission in Diabetes Patients    

__________________________________________________________________________________________________________________
**Group Members:**  
Aaron Hofman  
Arunachalesh M Kembhavimath  
Phone Pyae Zaw  
Jaztin Tabunda  

__________________________________________________________________________________________________________________

## Data Science Project Instructions – Master’s Program

Course Title: COMP 641 – Fundamentals of Data Science
Project Title:  End to End Data Science Project
Due Date: December 12, 2025

Overview
This project is a capstone-style assignment designed to apply the core concepts of data science covered throughout the course. You will identify a real-world problem, acquire and prepare data, conduct exploratory and statistical analysis, build and evaluate models, and present your findings in both technical and non-technical formats.

Objectives
By completing this project you will:
        Frame a real-world problem in data science terms.
        Collect, clean, and analyze data.
        Apply data science methods (EDA, modeling, evaluation).
        Communicate results effectively to stakeholders.
        Practice reproducibility and ethical data practices.

Proposal  Submission: (Due: Week 3)

Submit a one-page proposal including:
1. Project title and team members
2. Problem statement
3. Source of data (must be accessible)
4. Planned methods/techniques (e.g., regression, classification, clustering)
5. Ethical considerations

Project Components

Data Collection & Cleaning

    Students must acquire data from a reliable source, such as public datasets, APIs, or through web scraping techniques.
    Any challenges encountered during the data sourcing process should be clearly documented and explained.
    Students must clean and preprocess the data, including handling missing values and performing appropriate feature engineering steps.

Exploratory Data Analysis (EDA)

    Students are required to visualize data trends, identify outliers, and explore data distributions to gain insight into the dataset.
    At least three visualizations must be included, using tools such as matplotlib, seaborn, or similar libraries.
    A written narrative must accompany the visualizations, summarizing key findings from the exploratory analysis.

Modeling & Analysis

    Students must choose appropriate machine learning models based on the nature of their problem, such as regression, clustering, or classification.
    The dataset should be split into training and testing sets, or evaluated using cross-validation techniques.
    Model performance must be assessed using relevant evaluation metrics, such as accuracy, RMSE, or AUC.

Ethics & Bias Discussion

    Students must identify and discuss any potential biases present in the data collection process or modeling approach.
    The discussion should include the possible impact of these biases on stakeholders or end users of the system.

Presentation (Due: December 12)

Students are required to deliver a 5–10 minute in-class presentation summarizing their project. The presentation should clearly explain the problem addressed, the data used, the methods applied, and what are the key findings. Visual aids such as charts, graphs, or dashboards must be included to support the narrative. The presentation should be accessible to a broad audience and avoid excessive technical jargon.

 

Final Report (Due: December 12)

            Students must submit a technical report that is between 5 to 10 pages in length.
            The report should include an introduction and clearly framed problem statement. A detailed description of the data and methodology used must be provided in the report. Students must present and interpret the results of their analysis and modeling.


## Project Environment Preparation

### Project Imports:

In [6]:
# from ucimlrepo import fetch_ucirepo
import pandas as pd
import os
from tqdm.auto import tqdm
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

### Preparing CSV File paths for loading and Output Purposes:

This is just to set up the paths where we will be saving the encoded dataset for the model and the place where the csv file is

In [9]:
# Location of diabetic_data.csv.
# "<insert file path of csv file/dataset>"
import_df_location = "C:\\Users\\Jaztin T\\Downloads\\COMP-641-ProjectJupyterNotebook\\diabetic_data.csv"
# Locations for output files.
# "<insert file path of where you want output files to be>"
file_save_loaction = "C:\\Users\\Jaztin T\\Downloads\\COMP-641-ProjectJupyterNotebook"

# File location provided check.
if (import_df_location == ""):
    raise ValueError("import_df_location cannot be empty.")
if (file_save_loaction == ""):
    raise ValueError("file_save_loaction cannot be empty.")

### Setting up Variables for Output Text Information Files:

The following variables are used to produce the Output text files at the end of the notebook file

In [12]:
# Output text.
list_of_unique_and_nunique = ""
check_duplicates_in_first_two = ""
replacing_race_with_integers = ""
replacing_gender_with_integers = ""
replace_age_categories_with_integers = ""
removing_weight_column = ""
condensed_insurance_information = ""
replacing_medical_specialty_with_integers = ""
replacing_diagnoses_codes_with_integers = ""
replacing_medication_used_indication_with_integers = ""
replacing_max_glu_serum_with_integers = ""
replacing_A1Cresult_with_integers = ""
replacing_change_with_integers = ""
replacing_diabetesMed_with_integers = ""
replacing_readmitted_with_integers = ""

## Data Collection & Cleaning

### Data Loading 

The Main Diabetes Dataset is from diabetic_data.csv. This includes the information on hospital readmissions and patients.    

The other datasets are originally from the IDS_mapping.csv  file. These datasets have the meaning of some columns values in the Main Diabetes Dataset. They were split from the IDS_mapping.csv to their own .csv files. Then they were loaded in to their respective dataframes for ease of use later in the EDA stage.

In [16]:
# Importing datasets and loading them into their respective dataframes (Assuming it's in the same directory as jupyter notebook file)
# Main Diabetes dataset
DiabetesDF = pd.read_csv('diabetic_data.csv')

# Admission Source Type dataset. Derived from IDS_mapping.csv
# combined with the main diabetes dataset, one can tell the origin of the patient prior to hospital admission
admissionSourceDF = pd.read_csv('admission_source_types.csv')

# Admission Type dataset. Derived from IDS_mapping.csv
# combined with the main diabetes dataset, one can tell the patient's type of visit to the hospital (e.g. urgent, emergency, elective)
admissionTypeDF = pd.read_csv('admission_type.csv')

# Discharge disposition type dataset. Derived from IDS_mapping.csv
# combined with the main diabetes dataset, one can tell where the patient went after discharge from hospital
dischargeDispositionDF = pd.read_csv('discharge_disposition.csv')

# dictionary of dataframes in case we want to loop through each dataframe later
dataFrames = {
    "Diabetes Dataset": DiabetesDF,
    "Admission Source Dataset": admissionSourceDF,
    "Admission Type Dataset": admissionTypeDF,
    "Discharge Disposition Dataset": dischargeDispositionDF
}

### Data Initial Inspection - Seeing what we are dealing with

#### Summary Overview of Each Dataframe

In [19]:
# for each dataframe we have, we'll get a brief summary
for name, dataFrame in dataFrames.items():
    print(name + "----------------------------------------------\n")
    print(dataFrame.info())
    print("----------------------------------------------\n")

Diabetes Dataset----------------------------------------------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 


Looking at the information for each dataset, we find each dataset's data types, number of rows/entries, number of columns, and whether or not the datasets contain 'null' values. Firstly, looking at the diabetes dataset, one can see that the shape (rows x columns) of the dataset is 101766 rows/entries by 50 columns. Additionally, one can see that the dataset consists of integer values and strings/objects. Besides this, one may notice that most columns in the Diabetes dataset have non-null values in them. However, there are some columns that have null values. Namely, max_glu_serum and A1Cresult have a very low amount of non-null values in the dataset. To go into specifics, max_glu_serum has 5346 non-null values and A1Cresult has 17018 non-null values. Actions will most likely done on the columns later. Besides the main diabetes dataset, we have 'null' values in each dataset's "description" column. This doesn't seem problematic since it's only used for the mapping, so we don't really have to worry about those columns being null right now.   

#### Column/Feature Information

In [21]:
# for each dataframe we have, we'll point out some information about the columns 
# such as its number of features and the column names
for name, dataFrame in dataFrames.items():
    print(name + "----------------------------------------------\n")
    print(f"List of columns/features for {name}:\n {dataFrame.columns.tolist()} \n")
    print("Number of features:\n" + str(len(dataFrame.columns.tolist())) + "\n")
    print("----------------------------------------------\n")

Diabetes Dataset----------------------------------------------

List of columns/features for Diabetes Dataset:
 ['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'payer_code', 'medical_specialty', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'] 

Number of features:
50

--------------------------------------------

From looking at the information of each column, one can see the number of features in each dataset and what kind of information each dataset has.
However, one can see that we would need to merge the dataset to get a better idea of what categorical description corresponds to a patient's visit. 

#### First Five Elements of each Dataset

In [23]:
for name, dataFrame in dataFrames.items():
    print(name + "----------------------------------------------\n")
    print(f"{dataFrame.head(5)} \n")
    print("----------------------------------------------\n")

Diabetes Dataset----------------------------------------------

   encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0

Upon looking at the first five elements of each data, one can see the type of data we are dealing with, and some obvious issues. One obvious issue is the fact that the first 5 rows lack a weight value in the Diabetes dataset. In other words, the weight column should most likely be excluded since it doesn't look like we can get valuable information from it. Additionally, it shows that one would have to look out for "?" values in the dataset.

#### Missing value counts per column

In [26]:
for name, dataFrame in dataFrames.items():
    print(name + "'s Missing Value Counts Per Column:-------------------------------------------------------------\n")
    print(str(dataFrame.isna().sum()) + "\n")

Diabetes Dataset's Missing Value Counts Per Column:-------------------------------------------------------------

encounter_id                    0
patient_nbr                     0
race                            0
gender                          0
age                             0
weight                          0
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                      0
medical_specialty               0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                          0
diag_2                          0
diag_3                          0
number_diagnoses                0
max_glu_serum               96420
A1Cresult                   84748
metformin                       0
repaglinide                     0
na

From just glancing at the "missing value counts per column", as stated previously, one might come to the conclusion that only max_glu_serum (94% null) and A1Cresult (83% null) have missing values. However, we need to also consider that other columns in the diabetes dataset mark empty data cells with "?" mark. With that being said, merging the other datasets into the main diabetes datasets will be done to get a more accurate number of "null" values. Additionally, the count will include looking at rows with "?"

#### Looking for Unique Values

##### For Output Text Information File

In [30]:
# Method that prints both a list of the unique elements and the number of unique elements from a provided column name.
# The user can also state when this print method was used.
def unique_element_and_count_string(column_name, before_or_after_or_empty):
    string_to_return = f"Unique Elements {before_or_after_or_empty}: {str(DiabetesDF[column_name].unique())}\n"
    string_to_return += f"Unique Element Count {before_or_after_or_empty}: {str(DiabetesDF[column_name].nunique())}\n"
    return string_to_return

column_names_list = DiabetesDF.columns.tolist()
# Save list and count of unique elements from each column.
list_of_unique_and_nunique += "List and count of unique elements from each column in dataframe.\n"
for element in column_names_list:
    list_of_unique_and_nunique += f"Column Name: {element}\n"
    list_of_unique_and_nunique += unique_element_and_count_string(element, "")
    list_of_unique_and_nunique += "\n"

In [31]:
# Checking for duplicates in encounter_id and patient_nbr.
first_two_column_names = column_names_list[:2]
list_of_groups = []
for element in first_two_column_names:
    current_duplicate_check = DiabetesDF.groupby(element)
    list_of_groups.append(current_duplicate_check)
check_duplicates_in_first_two += "During earlier analysis, duplicate elements were found in patient_nbr.\n\n"

#### Making Adjustments to the Dataframes

##### Drop Weight Column

We drop the weight column because almost 97% of values are blank

In [35]:
# Drop 'weight' column, axis = 1 basically means drop the entire column. axis = 0 means drop the rows
DiabetesDF.drop('weight', axis=1, inplace=True)

##### For Output Text Information File

In [37]:
# Print drop reason.
removing_weight_column += "Dropped 'weight' column due to too many missing elements (97% missing).\n\n"

## EDA

## Modeling & Analysis

### Encoding - Preparing the dataset for the model

In [41]:
# Replace race text with integers.
replacing_race_with_integers += "Column Name: Race\n"
# Print replacement information.
replacing_race_with_integers += "Replacements:\n'?' --> 0\n'Caucasian' --> 1\n'AfricanAmerican' --> 2\n'Asian' --> 3\n'Hispanic' --> 4\n'Other' --> 5\n'Anything Else' --> -1\n"
# Print list of unique elements beforehand.
replacing_race_with_integers += unique_element_and_count_string("race", "before")
# Replacement function.
def replace_race_text(input_information):
    match input_information:
        case '?': # If '?' (not provided), return 0.
            return 0
        case 'Caucasian':
            return 1
        case 'AfricanAmerican':
            return 2
        case 'Asian':
            return 3
        case 'Hispanic':
            return 4
        case 'Other':
            return 5
        case _:
            return -1
# Perform replacement.
DiabetesDF['race'] = DiabetesDF['race'].apply(lambda x: replace_race_text(x))
# Print list of unique elements afterward to ensure replacement.
replacing_race_with_integers += unique_element_and_count_string("race", "after")
replacing_race_with_integers += "\n"

In [42]:
# Replace gender text with integers.
replacing_gender_with_integers += "Column Name: gender\n"
# Print replacement information.
replacing_gender_with_integers += "Replacements:\n'Unknown/Invalid' --> 0\n'Male' --> 1\n'Female' --> 2\n'Anything Else' --> -1\n"
# Print list of unique elements beforehand.
replacing_gender_with_integers += unique_element_and_count_string("gender", "before")
# Replacement function.
def replace_gender(input_information):
    match input_information:
        case 'Unknown/Invalid':
            return 0
        case 'Male':
            return 1
        case 'Female':
            return 2
        case _:
            return -1
# Perform replacement.
DiabetesDF['gender'] = DiabetesDF['gender'].apply(lambda x: replace_gender(x))
# Print list of unique elements afterward to ensure replacement.
replacing_gender_with_integers += unique_element_and_count_string("gender", "after")
replacing_gender_with_integers += "\n"

In [43]:
# Replace age text with integers.
replace_age_categories_with_integers += "Column Name: age\n"
# Print replacement information.
replace_age_categories_with_integers += "Replacements:\n'[0-10)' --> 0\n'[10-20)' --> 1\n'[20-30)' --> 2\n'[30-40)' --> 3\n'[40-50)' --> 4\n'[50-60)' --> 5\n'[60-70)' --> 6\n'[70-80)' --> 7\n'[80-90)' --> 8\n'[90-100)' --> 9\n'Anything Else' --> -1\n"
# Print list of unique elements beforehand.
replace_age_categories_with_integers += unique_element_and_count_string("age", "before")
# Replacement function.
def replace_age_ranges(input_information):
    match input_information:
        case '[0-10)':
            return 0
        case '[10-20)':
            return 1
        case '[20-30)':
            return 2
        case '[30-40)':
            return 3
        case '[40-50)':
            return 4
        case '[50-60)':
            return 5
        case '[60-70)':
            return 6
        case '[70-80)':
            return 7
        case '[80-90)':
            return 8
        case '[90-100)':
            return 9
        case _:
            return -1
# Perform replacement.
DiabetesDF['age'] = DiabetesDF['age'].apply(lambda x: replace_age_ranges(x))
# Print list of unique elements afterward to ensure replacement.
replace_age_categories_with_integers += unique_element_and_count_string("age", "after")
replace_age_categories_with_integers += "\n"

In [44]:
# Condense insurance information.
condensed_insurance_information += "Column Name: payer_code\n"
# Print condensing information.
condensed_insurance_information += "Condensing:\n'?' --> 0\n'SP' --> 1\n'Anything Else' --> 2\n"
# Print reason for condensing.
condensed_insurance_information += "Reason for condensing:\nPatients who don't use insurance are less likely to return for additional treatment, which puts them at increased risk for further injury.\n"
# Print list of unique elements beforehand.
condensed_insurance_information += unique_element_and_count_string("payer_code", "before")
# Filtering function.
def filter_payer_code(input_information):
    # Insurance information not provided.
    if (input_information == '?'):
        return 0
    # Patient did not use insurance. I don't know if 'SP' stands for 'self-pay' or not.
    elif (input_information == 'SP'):
        return 1
    # Patient used insurance.
    else:
        return 2
# Perform the filtering operation.
DiabetesDF['payer_code'] = DiabetesDF['payer_code'].apply(lambda x: filter_payer_code(x))
# Print list of unique elements afterward to ensure filtering.
condensed_insurance_information += unique_element_and_count_string("payer_code", "after")
condensed_insurance_information += "\n"

In [45]:
# Replace the elements in a given column with integers.
def replace_elements_with_ints(column_name):
    return_question_mark_associated_element = 0

    # Get the list of unique column elements from the column in question.
    unique_column_elements = DiabetesDF[column_name].unique().tolist()

    # Check if a '?' element exists, and move it to the front of the list of unique column elements if it does.
    try:
        question_mark_location = unique_column_elements.index('?')
        unique_column_elements.pop(question_mark_location)
        unique_column_elements.insert(0, '?')

        # Additional information needs to be printed
        if (question_mark_location == 0):
            return_question_mark_associated_element = 1
        if (question_mark_location > 0):
            return_question_mark_associated_element = 2
    except:
        pass

    # Convert back to numpy array for printing purposes.
    unique_column_elements = np.array(unique_column_elements)

    # Get the number of unique column elements from the column in question, and then generate a list of elements from 0 up to one less than the number of unique elements.
    unique_column_element_count = DiabetesDF[column_name].nunique()
    list_of_numbers = list(range(0, unique_column_element_count))

    # Pair each number to a unique column element and perform the replacement.
    unique_column_elements_numbers_pairing = {key: element for key, element in zip(unique_column_elements, list_of_numbers)}
    DiabetesDF[column_name] = DiabetesDF[column_name].apply(lambda x: unique_column_elements_numbers_pairing[x])

    # If a '?' element exists, the temporary reordering of unique elements was done, so additional information needs to be printed.
    if (return_question_mark_associated_element == 1):
        return f"The element associated with '?' is '{unique_column_elements_numbers_pairing['?']}'\n"
    elif (return_question_mark_associated_element == 2):
        return (
            f"Temporary reordering of unique elements: {unique_column_elements}\n" +
            f"The element associated with '?' is '{unique_column_elements_numbers_pairing['?']}'\n"
            )
    else:
        return ""
# Replacement information.
replace_elements_with_ints_string = "Replacement:\nEverything --> After moving '?' (if it exists) to the front of the unique elements list, each element is assigned a unique number, starting from zero up to one less than the number of unique elements.\n"


In [46]:
# Replacing medical specialty strings with integers.
# Despite 53% of element missing, the medical speciality examining a patient is very important.
#  Example: Emergency/Trauma is much more important than Dentistry.
def consensing_medical_specialties(input_information):
    match input_information:
        # Primary Care
        case x if (x == 'Family/GeneralPractice') or (x == 'InternalMedicine') or (x == 'Pediatrics') or (x == 'Hospitalist') or (x == 'Osteopath') or (x == 'Resident') or (x == 'PhysicianNotFound'):
            return 'Primary Care'
        # Cardiovascular System - Non-surgery.
        case x if (x == 'Cardiology') or (x == 'Cardiology-Pediatric'):
            return 'Cardiovascular System - Non-surgery'
        # Neurology - Non-surgery.
        case x if (x == 'Neurology') or (x == 'Neurophysiology') or (x == 'Pediatrics-Neurology'):
            return 'Neurology - Non-surgery'
        # Psychiatry
        case x if (x == 'Psychiatry') or (x == 'Psychiatry-Child/Adolescent') or (x == 'Psychiatry-Addictive') or (x == 'Psychology'):
            return 'Psychiatry'
        case _:
            return input_information
# Condense some of the related medical specialities into groups.
replacing_medical_specialty_with_integers += "Column Name: medical_specialty\n"
# Perform consensing.
DiabetesDF['medical_specialty'] = DiabetesDF['medical_specialty'].apply(lambda x: consensing_medical_specialties(x))
# Print list and number of unique elements after condensing.
replacing_medical_specialty_with_integers += unique_element_and_count_string('medical_specialty', "After Condensing")
# Print notes.
replacing_medical_specialty_with_integers += ("Notes:\nDespite 53% of elements missing, the medical speciality examining a patient is very important.\n" +
                                              "Example:\nEmergency/Trauma is much more important than Dentistry.\n" + 
                                              "In addition, 'DCPTEAM' possibly stands for 'Dynamic Care Planning Team'.\n")
# Print replacement message.
replacing_medical_specialty_with_integers += replace_elements_with_ints_string
# Print list and number of unique elements beforehand.
replacing_medical_specialty_with_integers += unique_element_and_count_string("medical_specialty", "before")
# Perform replacement.
replacing_medical_specialty_with_integers += replace_elements_with_ints('medical_specialty')
# Print list and number of unique elements afterward to ensure replacement.
replacing_medical_specialty_with_integers += unique_element_and_count_string("medical_specialty", "after")
replacing_medical_specialty_with_integers += "\n"

In [47]:
# Replace the codes for diagnoses 1, 2, and 3 with integers.
for i in range(18, 21):
    current_column_name = column_names_list[i]
    # Print column name.
    replacing_diagnoses_codes_with_integers += f"Column Name: {current_column_name}\n"
    # Print reason for replacement.
    replacing_diagnoses_codes_with_integers += replace_elements_with_ints_string
    # Print list and number of unique elements beforehand.
    replacing_diagnoses_codes_with_integers += unique_element_and_count_string(current_column_name, "before")
    # Perform replacement.
    replacing_diagnoses_codes_with_integers += replace_elements_with_ints(current_column_name)
    # Print list and number of unique elements afterward to ensure replacement.
    replacing_diagnoses_codes_with_integers += unique_element_and_count_string(current_column_name, "after")
    replacing_diagnoses_codes_with_integers += "\n"


In [48]:
# Replace no, steady, up, and down with corresponding integers.
def replace_no_steady_up_down(input_information):
    match input_information.lower():
        case 'no':
            return 0
        case 'steady':
            return 1
        case 'up':
            return 2
        case 'down':
            return 3
        case _:
            return -1
# Print medication note.
replacing_medication_used_indication_with_integers += "Medication Note:\nWhile some medications were not administered, medications are still important and, as a result, they weren't dropped.\n\n"
for i in range(24, 47):
    current_column_name = column_names_list[i]
    # Print column name.
    replacing_medication_used_indication_with_integers += f"Column Name: {current_column_name}\n"
    # Print reason for replacement.
    replacing_medication_used_indication_with_integers += "Replacements:\n'no' --> 0\n'steady' --> 1\n'up' --> 2\n'down' --> 3\n'Anything Else' --> -1\n"
    # Print list and number of unique elements beforehand.
    replacing_medication_used_indication_with_integers += unique_element_and_count_string(current_column_name, "before")
    # Perform replacement.
    DiabetesDF[current_column_name] = DiabetesDF[current_column_name].apply(lambda x: replace_no_steady_up_down(x))
    # Print list and number of unique elements afterward to ensure replacement.
    replacing_medication_used_indication_with_integers += unique_element_and_count_string(current_column_name, "after")
    replacing_medication_used_indication_with_integers += "\n"

In [49]:
# Replace elements in 'max_glu_serum' column.
replacing_max_glu_serum_with_integers += "Column Name: max_glu_serum\n"
# Print replacement information.
replacing_max_glu_serum_with_integers += "Replacements:\n'nan' --> 0\n'norm' --> 1\n'>200' --> 2\n'>300' --> 3\n'Anything Else' --> -1\n"
# Print note.
replacing_max_glu_serum_with_integers += "Note:\n'None' in the excel sheet might be 'nan' in Pandas. That explains the jump from 3 to 4 in the number of unique elements.\n"
# Print list and number of unique elements beforehand.
replacing_max_glu_serum_with_integers += unique_element_and_count_string("max_glu_serum", "before")
# Replace elements with corresponding integers.
def replace_max_glu_serum_elements(input_information):
    match str(input_information).lower():
        case 'nan':
            return 0
        case 'norm':
            return 1
        case '>200':
            return 2
        case '>300':
            return 3
        case _:
            return -1
# Perform replacement.
DiabetesDF['max_glu_serum'] = DiabetesDF['max_glu_serum'].apply(lambda x: replace_max_glu_serum_elements(x))
# Print list and number of unique elements afterward to ensure replacement.
replacing_max_glu_serum_with_integers += unique_element_and_count_string("max_glu_serum", "after")
replacing_max_glu_serum_with_integers += "\n"

In [50]:
# Replace elements in 'A1Cresult' column.
replacing_A1Cresult_with_integers += "Column Name: A1Cresult\n"
# Perform replacement information.
replacing_A1Cresult_with_integers += "Replacements:\n'nan' --> 0\n'norm' --> 1\n'>7' --> 2\n'>8' --> 3\n'Anything Else' --> -1\n"
# Print note.
replacing_A1Cresult_with_integers += "Note:\n'None' in the excel sheet might be 'nan' in Pandas. That explains the jump from 3 to 4 in the number of unique elements.\n"
# Print list and number of unique elements beforehand.
replacing_A1Cresult_with_integers += unique_element_and_count_string("A1Cresult", "before")
# Replace elements with corresponding integers.
def replace_A1Cresult_elements(input_information):
    match str(input_information).lower():
        case 'nan':
            return 0
        case 'norm':
            return 1
        case '>7':
            return 2
        case '>8':
            return 3
        case _:
            return -1
# Perform replacement.
DiabetesDF['A1Cresult'] = DiabetesDF['A1Cresult'].apply(lambda x: replace_A1Cresult_elements(x))
# Print list and number of unique elements afterward to ensure replacement.
replacing_A1Cresult_with_integers += unique_element_and_count_string("A1Cresult", "after")
replacing_A1Cresult_with_integers += "\n"

In [51]:
# Replace elements in 'change' column.
replacing_change_with_integers += "Column Name: change\n"
# Print replacement information.
replacing_change_with_integers += "Replacements:\n'no' --> 0\n'ch' --> 1\n'Anything Else' --> -1\n"
# Print list and number of unique elements beforehand.
replacing_change_with_integers += unique_element_and_count_string("change", "before")
# Replace elements with corresponding integers.
def replace_change_elements(input_information):
    match str(input_information).lower():
        case 'no':
            return 0
        case 'ch':
            return 1
        case _:
            return -1
# Perform replacement.
DiabetesDF['change'] = DiabetesDF['change'].apply(lambda x: replace_change_elements(x))
# Print list and number of unique elements afterward to ensure replacement.
replacing_change_with_integers += unique_element_and_count_string("change", "after")
replacing_change_with_integers += "\n"


In [52]:
# Replace elements in 'diabetesMed' column.
replacing_diabetesMed_with_integers += "Column Name: diabetesMed\n"
# Print replacement information.
replacing_diabetesMed_with_integers += "Replacements:\n'no' --> 0\n'yes' --> 1\n'Anything Else' --> -1\n"
# Print list and number of unique elements beforehand.
replacing_diabetesMed_with_integers += unique_element_and_count_string("diabetesMed", "before")
# Replace elements with corresponding integers.
def replace_diabetesMed_elements(input_information):
    match str(input_information).lower():
        case 'no':
            return 0
        case 'yes':
            return 1
        case _:
            return -1
# Perform replacement.
DiabetesDF['diabetesMed'] = DiabetesDF['diabetesMed'].apply(lambda x: replace_diabetesMed_elements(x))
# Print list and number of unique elements afterward to ensure replacement.
replacing_diabetesMed_with_integers += unique_element_and_count_string("diabetesMed", "after")
replacing_diabetesMed_with_integers += "\n"


In [53]:
# Replace elements in 'readmitted' column.
replacing_readmitted_with_integers += "Column Name: readmitted\n"
# Print replacement information.
replacing_readmitted_with_integers += "Replacements:\n'no' --> 0\n'<30' --> 1\n'>30' --> 2\n'Anything Else' --> -1\n"
# Print list and number of unique elements beforehand.
replacing_readmitted_with_integers += unique_element_and_count_string("readmitted", "before")
# Replace elements with corresponding integers.
def replace_readmitted_elements(input_information):
    match str(input_information).lower():
        case 'no':
            return 0
        case '<30':
            return 1
        case '>30':
            return 2
        case _:
            return -1
# Perform replacement.
DiabetesDF['readmitted'] = DiabetesDF['readmitted'].apply(lambda x: replace_readmitted_elements(x))
# Print list and number of unique elements afterward to ensure replacement.
replacing_readmitted_with_integers += unique_element_and_count_string("readmitted", "after")
replacing_readmitted_with_integers += "\n"


## Creating the Encoded CSV Files and Text Information Files

In [55]:
# print("Outputting information to text files.")
# final_output_string_1 = (
#     "Below, please find a list of the unique elements in each column along with how many unique elements there are.\n\n\n"
#     f"{list_of_unique_and_nunique}\n\n"
# )
# final_output_string_2 = (
#     "Introductory Info:\n"
#     "This file: Information on the modifications done to the diabetic_data.csv file and why they were done.\n"
#     "unique_information.txt: A list of unique elements in each column along with how many unique elements there are.\n"
#     "output.csv: A version of diabetic_data.csv modified with the edits described in this file.\n"
#     "output_grouped.csv: A version of output.csv where everything has been grouped together based on patient_nbr.\n"
#     "Percentages for the amount of missing elements for some of the columns can be found here: https://onlinelibrary.wiley.com/doi/10.1155/2014/781670\n\n\n\n\n"
#     "Modifications to diabetic_data.csv:\n\n\n"
#     f"{check_duplicates_in_first_two}\n\n"
#     f"{replacing_race_with_integers}\n\n"
#     f"{replacing_gender_with_integers}\n\n"
#     f"{replace_age_categories_with_integers}\n\n"
#     f"{removing_weight_column}\n\n"
#     f"{condensed_insurance_information}\n\n"
#     f"{replacing_medical_specialty_with_integers}\n\n"
#     f"{replacing_diagnoses_codes_with_integers}\n\n"
#     f"{replacing_medication_used_indication_with_integers}\n\n"
#     f"{replacing_max_glu_serum_with_integers}\n\n"
#     f"{replacing_A1Cresult_with_integers}\n\n"
#     f"{replacing_change_with_integers}\n\n"
#     f"{replacing_diabetesMed_with_integers}\n\n"
#     f"{replacing_readmitted_with_integers}"
# )

In [56]:
# # Save unique element and count information to text file.
# with open("unique_information.txt", 'w') as file:
#     file.write(final_output_string_1)
# # Save explanation information to text file.
# with open("explanation_information.txt", 'w') as file:
#     file.write(final_output_string_2)
# print("Information outputted to text file.")

In [57]:
# # Create modified csv file from resulting dataframe.
# print("Outputting modified csv file.")
# DiabetesDF.to_csv(os.path.join(file_save_loaction, 'output.csv'), index=False)
# print("Modified csv file outputted.")

In [58]:
# # Group resulting dataframe by the 'patient_nbr' column.
# print("Starting Grouping")
# tqdm.pandas()
# # https://stackoverflow.com/questions/60963409/creating-a-dictionary-of-dictionaries-from-groupby
# DiabetesDF_grouped = DiabetesDF.groupby('patient_nbr')
# group_of_groups = DiabetesDF_grouped.progress_apply(lambda x: x.set_index('encounter_id').to_dict())
# print("Grouping Done")

In [59]:
# # Create grouped csv file from the grouped dataframe.
# print("Outputting grouped version of modified csv file.")
# group_of_groups.to_csv(os.path.join(file_save_loaction, 'output_grouped.csv'), index=False)
# print("Grouped version of modified csv file outputted.")